src/System.Linq/src/System/Linq/Enumerable.cs

```diff
@@ -447,14 +603,10 @@ public override bool MoveNext()
                         state = 2;
                         goto case 2;
                     case 2:
-                        while (_enumerator.MoveNext())
+                        if (_enumerator.MoveNext())
```

Comment:
No need. List<T>.Enumerator throws to prohibit code like:

```C#
foreach(var item in list)
{
  // alter list here and don't jump out of the loop
}

And `Select_SourceListGetsModifiedDuringIteration_ExceptionIsPropagated` insists that we throw on:

```C#
foreach(var item in list.Select(λ))
{
  // alter list here and don't jump out of the loop
}

`ToArray` doesn't allow for that possibility, so iterating is a valid optimisation.

(There's the potential for problems if the source list to be changed during the call by another thread, but that involves getting non-threadsafe behaviour from methods that don't claim to be threadsafe, and could fail either way).